torchdata/nodes/_populate_queue.py

```diff
@@ -0,0 +1,40 @@
+# pyre-unsafe
+import queue
+import threading
+from typing import Iterable
+
+from torchdata.nodes.exception_wrapper import ExceptionWrapper, StartupExceptionWrapper
+
+
+def _populate_queue(
+    source: Iterable,
+    q: queue.Queue,
+    semaphore: threading.BoundedSemaphore,
+    stop_event: threading.Event,
+):
+    """Note that this is only intended to be used
+    by a single thread at once. Each instance creates its own iter for source so
+    if this is called with multiple threads, you may get duplicates if
+    source is not sharded properly.
+    """
```

Comment:
Basic question. In the canonical `DataLoader` we have one index queue per worker process, and one output queue per worker. In the new setup, would we have one input queue and one output queue per instance of `ParallelMapper` (correct me if I confused myself here). How many `ParallelMapper`s would we have in a typical setup ? I'd assume >1 because otherwise there is only one thread doing dataset access.

Reply:
for the torch.utils.data.DataLoader, there is one output queue shared by all workers, but otherwise you're right, there's one index queue per worker. The reason they do it that way is to give guarantees on determinism, so that work is always assigned round-robin style to the workers.

Currently we have just one index queue but all workers can pull from it, so parallelism isn't constrained, but we might want to adopt something similar to torch.utils.data.DataLoader to guarantee things like random transforms and state. We can do that in a follow up task